From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Roch Blonndiaux <contact@roch-blondiaux.com>
Date: Mon, 18 Sep 2023 16:23:00 +0200
Subject: [PATCH] feat: versioning update


diff --git a/build.gradle.kts b/build.gradle.kts
index 1c1dd088187f5700e3e4b7ec3b4b6c0acd5c119a..93f428b56fd75ccb5e6f16cdbdb3e5b8cfb98e6e 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -13,7 +13,7 @@ configurations.named(log4jPlugins.compileClasspathConfigurationName) {
 val alsoShade: Configuration by configurations.creating
 
 dependencies {
-    implementation(project(":pufferfish-api")) // Pufferfish // Paper
+    implementation(project(":castaway-api")) // Pufferfish // Paper
     // Pufferfish start
     implementation("io.papermc.paper:paper-mojangapi:1.19.2-R0.1-SNAPSHOT") {
         exclude("io.papermc.paper", "paper-api")
diff --git a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
index 80553face9c70c2a3d897681e7761df85b22d464..5fa91e07b407a0c75011de36f260d0538c6c1e05 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
@@ -11,7 +11,7 @@ public final class Versioning {
     public static String getBukkitVersion() {
         String result = "Unknown-Version";
 
-        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/gg.pufferfish.pufferfish/pufferfish-api/pom.properties"); // Pufferfish
+        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/gg.pufferfish.pufferfish/castaway-api/pom.properties"); // Pufferfish
         Properties properties = new Properties();
 
         if (stream != null) {
